src/test/java/io/cdap/plugin/batch/source/BSONObjectToRecordTransformerTest.java

```diff
@@ -0,0 +1,184 @@
+                                    Schema.Field.of("string_field", Schema.nullableOf(Schema.of(Schema.Type.STRING))),
+                                    Schema.Field.of("boolean_field", Schema.nullableOf(Schema.of(Schema.Type.BOOLEAN))),
+                                    Schema.Field.of("bytes_field", Schema.nullableOf(Schema.of(Schema.Type.BYTES))),
+                                    Schema.Field.of("null_field", Schema.nullableOf(Schema.of(Schema.Type.STRING))),
+                                    Schema.Field.of("array_field", Schema.nullableOf(
+                                      Schema.arrayOf(Schema.nullableOf(Schema.of(Schema.Type.LONG))))),
+                                    Schema.Field.of("nested_object", Schema.nullableOf(nestedRecordSchema)),
+                                    Schema.Field.of("object_to_map", Schema.nullableOf(
+                                      Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING)))));
+
+    BSONObject bsonObject = BasicDBObjectBuilder.start().get();
+    BSONObjectToRecordTransformer transformer = new BSONObjectToRecordTransformer(schema);
+    StructuredRecord transformed = transformer.transform(bsonObject);
+    Assert.assertNull(transformed.get("int_field"));
+    Assert.assertNull(transformed.get("long_field"));
+    Assert.assertNull(transformed.get("double_field"));
+    Assert.assertNull(transformed.get("string_field"));
+    Assert.assertNull(transformed.get("boolean_field"));
+    Assert.assertNull(transformed.get("bytes_field"));
+    Assert.assertNull(bsonObject.get("null_field"));
+    Assert.assertNull(transformed.get("array_field"));
+    Assert.assertNull(transformed.get("nested_object"));
+    Assert.assertNull(transformed.get("object_to_map"));
+  }
+
+  @Test
+  public void testTransformUnexpectedFormat() {
+    Schema schema = Schema.recordOf("schema", Schema.Field.of("union_field", Schema.unionOf(
+      Schema.of(Schema.Type.LONG),
+      Schema.of(Schema.Type.STRING)))
+    );
+
+    BSONObject bsonObject = BasicDBObjectBuilder.start()
+      .add("union_field", 2019L)
+      .get();
+
+    thrown.expect(UnexpectedFormatException.class);
+    BSONObjectToRecordTransformer transformer = new BSONObjectToRecordTransformer(schema);
+    transformer.transform(bsonObject);
+  }
+
+  @Test
+  public void testTransformInvalidMapSchema() {
+    Schema schema = Schema.recordOf("schema", Schema.Field.of("invalid_key_type", Schema.mapOf(
+      Schema.of(Schema.Type.LONG), Schema.of(Schema.Type.STRING))));
+
+    BSONObject mapObject = BasicDBObjectBuilder.start()
+      .add("key", "value")
+      .get();
+
+    BSONObject bsonObject = BasicDBObjectBuilder.start()
+      .add("invalid_key_type", mapObject)
+      .get();
+
+    thrown.expect(UnexpectedFormatException.class);
+    BSONObjectToRecordTransformer transformer = new BSONObjectToRecordTransformer(schema);
+    transformer.transform(bsonObject);
+  }
+}
```

Comment:
can you add some tests for when the value in the BSONObject does not match the type expected in the Schema.

Reply:
[Done](https://github.com/data-integrations/mongodb-plugins/pull/3/files#diff-19f4357d1cb90adb8abc786cce615966R209).